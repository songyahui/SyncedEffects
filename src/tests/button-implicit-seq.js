"use hopscript"

function plus (x, y) { return x+y };

var hh = require("hiphop");

var sigIn={accessibility: hh.IN};
var sigOut={accessibility: hh.OUT};

var BUTTON =
    <hh.module UL=${sigIn} UR=${sigIn} LL=${sigIn} LR=${sigIn}
	       WATCH_MODE_COMMAND=${sigOut}
	       ENTER_SET_WATCH_MODE_COMMAND=${sigOut}
	       SET_WATCH_COMMAND=${sigOut}
	       NEXT_WATCH_TIME_POSITION_COMMAND=${sigOut}
	       EXIT_SET_WATCH_MODE_COMMAND=${sigOut}
	       TOGGLE_24H_MODE_COMMAND=${sigOut}
	       TOGGLE_CHIME_COMMAND=${sigOut}
	       STOPWATCH_MODE_COMMAND=${sigOut}
	       START_STOP_COMMAND=${sigOut}
	       LAP_COMMAND=${sigOut}
	       ALARM_MODE_COMMAND=${sigOut}
	       ENTER_SET_ALARM_MODE_COMMAND=${sigOut}
	       SET_ALARM_COMMAND=${sigOut}
	       NEXT_ALARM_TIME_POSITION_COMMAND=${sigOut}
	       TOGGLE_ALARM_COMMAND=${sigOut}
	       STOP_ALARM_BEEP_COMMAND=${sigOut}
	       EXIT_SET_ALARM_MODE_COMMAND=${sigOut}>

    <!-- global loop -->
      <hh.parallel>
	<hh.loop>
    <!-- watch / set-watch mode -->
	  <hh.emit WATCH_MODE_COMMAND/>
	  <hh.trap WATCH_AND_SET_WATCH_MODE>
    <!-- watch mode -->
            <hh.loop>
	      <hh.abort UL>
		<hh.parallel>
                  <hh.sequence>
		    <hh.await LL/>
		    <hh.exit WATCH_AND_SET_WATCH_MODE/>
                  </hh.sequence>
                  <hh.loop>
		    <hh.await LR/>
		    <hh.emit TOGGLE_24H_MODE_COMMAND/>
		  </hh.loop>
		</hh.parallel>
	      </hh.abort>
<!-- set-watch mode -->
	      <hh.emit ENTER_SET_WATCH_MODE_COMMAND/>
	      <hh.abort UL>
		<hh.parallel>
                  <hh.loop>
		    <hh.await LL/>
		    <hh.emit NEXT_WATCH_TIME_POSITION_COMMAND/>
		  </hh.loop>
		  <hh.loop>
		    <hh.await LR/>
		    <hh.emit SET_WATCH_COMMAND/>
		  </hh.loop>
		</hh.parallel>
	      </hh.abort>
	      <hh.emit EXIT_SET_WATCH_MODE_COMMAND/>
	    </hh.loop>
	  </hh.trap>
<!-- stopwatch mode -->
	  <hh.emit STOPWATCH_MODE_COMMAND/>
	  <hh.abort LL>
            <hh.parallel>
	      <hh.loop>
		<hh.await LR/>
		<hh.emit START_STOP_COMMAND/>
	      </hh.loop>
	      <hh.loop>
		<hh.await UR/>
		<hh.emit LAP_COMMAND/>
	      </hh.loop>
            </hh.parallel>
	  </hh.abort>
<!-- alarm / set alarm mode -->
	  <hh.emit ALARM_MODE_COMMAND/>
	  <hh.trap ALARM_AND_SET_ALARM_MODE>
            <hh.loop>
<!-- alarm mode -->
	      <hh.abort UL>
		<hh.parallel>
		  <hh.parallel>
		    <hh.sequence>
		      <hh.await LL/>
		      <hh.exit ALARM_AND_SET_ALARM_MODE/>
		    </hh.sequence>
		    <hh.loop>
		      <hh.await LR/>
		      <hh.emit TOGGLE_CHIME_COMMAND/>
		    </hh.loop>
		  </hh.parallel>
		  <hh.loop>
		    <hh.await UR/>
		    <hh.emit TOGGLE_ALARM_COMMAND/>
		  </hh.loop>
		</hh.parallel>
	      </hh.abort>
<!-- set-alarm mode -->
	      <hh.emit ENTER_SET_ALARM_MODE_COMMAND/>
	      <hh.abort UL>
		<hh.parallel>
                  <hh.loop>
		    <hh.await LL/>
		    <hh.emit NEXT_ALARM_TIME_POSITION_COMMAND/>
		  </hh.loop>
		  <hh.loop>
		    <hh.await LR/>
		    <hh.emit SET_ALARM_COMMAND/>
		  </hh.loop>
		</hh.parallel>
	      </hh.abort>
	      <hh.emit EXIT_SET_ALARM_MODE_COMMAND/>
	    </hh.loop>
	  </hh.trap>
	</hh.loop>
	<hh.every UR>
	  <hh.emit STOP_ALARM_BEEP_COMMAND />
	</hh.every>
      </hh.parallel>
    </hh.module>;

exports.prg = new hh.ReactiveMachine(BUTTON, "BUTTON");
