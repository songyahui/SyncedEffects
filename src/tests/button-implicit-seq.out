BUTTON> ;
--- Output: WATCH_MODE_COMMAND
BUTTON> LR;
--- Output: TOGGLE_24H_MODE_COMMAND
BUTTON> LR;
--- Output: TOGGLE_24H_MODE_COMMAND
BUTTON> ;
--- Output:
BUTTON> UL;
--- Output: ENTER_SET_WATCH_MODE_COMMAND
BUTTON> LL;
--- Output: NEXT_WATCH_TIME_POSITION_COMMAND
BUTTON> LR;
--- Output: SET_WATCH_COMMAND
BUTTON> LL;
--- Output: NEXT_WATCH_TIME_POSITION_COMMAND
BUTTON> UL;
--- Output: EXIT_SET_WATCH_MODE_COMMAND
BUTTON> LR;
--- Output: TOGGLE_24H_MODE_COMMAND
BUTTON> ;
--- Output:
BUTTON> LL;
--- Output: STOPWATCH_MODE_COMMAND
BUTTON> ;
--- Output:
BUTTON> LR;
--- Output: START_STOP_COMMAND
BUTTON> UR;
--- Output: LAP_COMMAND STOP_ALARM_BEEP_COMMAND
BUTTON> LR;
--- Output: START_STOP_COMMAND
BUTTON> ;
--- Output:
BUTTON> LL;
--- Output: ALARM_MODE_COMMAND
BUTTON> ;
--- Output:
BUTTON> LR;
--- Output: TOGGLE_CHIME_COMMAND
BUTTON> UR;
--- Output: TOGGLE_ALARM_COMMAND STOP_ALARM_BEEP_COMMAND
BUTTON> UL;
--- Output: ENTER_SET_ALARM_MODE_COMMAND
BUTTON> LR;
--- Output: SET_ALARM_COMMAND
BUTTON> LL;
--- Output: NEXT_ALARM_TIME_POSITION_COMMAND
BUTTON> LR;
--- Output: SET_ALARM_COMMAND
BUTTON> UL;
--- Output: EXIT_SET_ALARM_MODE_COMMAND
BUTTON> LR;
--- Output: TOGGLE_CHIME_COMMAND
BUTTON> UR;
--- Output: TOGGLE_ALARM_COMMAND STOP_ALARM_BEEP_COMMAND
BUTTON> ;
--- Output:
BUTTON> LL;
--- Output: WATCH_MODE_COMMAND
BUTTON> ;
--- Output:
